File: src/lib/backup.ts
import type {
  ExportPayload,
  ProductObservation,
  ProductSnapshot,
  ShoppingList,
  ShoppingListItem,
} from "@/types/domain";
import { validateProductObservationInput } from "@/lib/validators";

const BACKUP_VERSION = "1.0.0";

export function serializeShoppingBackup(payload: {
  lists: ShoppingList[];
  items: ShoppingListItem[];
  snapshots: ProductSnapshot[];
  observations?: ProductObservation[];
}): ExportPayload {
  return {
    version: BACKUP_VERSION,
    createdAt: Date.now(),
    lists: payload.lists,
    items: payload.items,
    snapshots: payload.snapshots,
    observations: payload.observations ?? [],
  };
}

export function serializeShoppingBackupToJson(payload: ExportPayload): string {
  return JSON.stringify(payload, null, 2);
}

export function parseShoppingBackup(raw: string): ExportPayload {
  const parsed = JSON.parse(raw) as ExportPayload;

  if (parsed.version !== BACKUP_VERSION) {
    throw new Error("unsupported backup version");
  }

  if (!Array.isArray(parsed.lists) || !Array.isArray(parsed.items) || !Array.isArray(parsed.snapshots)) {
    throw new Error("invalid backup payload");
  }

  if (parsed.observations !== undefined && !Array.isArray(parsed.observations)) {
    throw new Error("invalid backup observations");
  }

  if (
    parsed.observations?.some((observation) =>
      !validateProductObservationInput(observation).valid,
    )
  ) {
    throw new Error("invalid backup observation record");
  }

  return parsed;
}

File: app/src-tauri/src/backup.rs
use serde_json::Value;

fn ensure_arrays(payload: &Value) -> bool {
  payload.get("lists").is_some_and(Value::is_array)
    && payload.get("items").is_some_and(Value::is_array)
    && payload.get("snapshots").is_some_and(Value::is_array)
}

fn ensure_version(payload: &Value) -> bool {
  payload.get("version").and_then(Value::as_str) == Some("1.0.0")
}

pub fn validate_payload(raw: &str) -> Result<Value, String> {
  let payload: Value = serde_json::from_str(raw)
    .map_err(|error| format!("invalid backup JSON: {error}"))?;

  if !ensure_version(&payload) {
    return Err("unsupported backup version".to_string());
  }

  if !ensure_arrays(&payload) {
    return Err("backup payload missing required lists/items/snapshots".to_string());
  }

  Ok(payload)
}
